File: java/src/main/java/com/blamedcloud/parsertongue/tokenizer/Tokenizer.java
package com.blamedcloud.parsertongue.tokenizer;

import static com.blamedcloud.parsertongue.tokenizer.DefaultGrammarConstants.EMPTY_NAME;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Tokenizer {

    private List<Token> tokens;
    private int index;
    private boolean exhausted;
    private TokenizerTypeList tokenizerTypeList;
    private boolean ignoreWhiteSpace;
    private final Pattern whiteSpacePattern = Pattern.compile("\\s+");

    public Tokenizer() {
        setDefaults();
        tokenizerTypeList = TokenizerTypeList.defaultGrammarTTL();
        ignoreWhiteSpace = true;
    }

    public Tokenizer(TokenizerTypeList ttl) {
        setDefaults();
        tokenizerTypeList = ttl;
        ignoreWhiteSpace = true;
    }

    public Tokenizer(boolean ignoreWS) {
        setDefaults();
        tokenizerTypeList = TokenizerTypeList.defaultGrammarTTL();
        ignoreWhiteSpace = ignoreWS;
    }

    public Tokenizer(TokenizerTypeList ttl, boolean ignoreWS) {
        setDefaults();
        tokenizerTypeList = ttl;
        ignoreWhiteSpace = ignoreWS;
    }

    private void setDefaults() {
        tokens = new ArrayList<>();
        index = 0;
        exhausted = false;
    }

    private void resetFromTokenList(List<Token> tokenList) {
        setDefaults();
        tokens = tokenList;
    }

    public boolean hasSameTokens(Tokenizer other) {
        if (tokens.size() != other.tokens.size()) {
            return false;
        }
        for (int i = 0; i < tokens.size(); i++) {
            if (!tokens.get(i).isSameAs(other.tokens.get(i))) {
                return false;
            }
        }
        return true;
    }

    public int size() {
        return tokens.size();
    }

    public boolean nextToken() {
        if (index + 1 == size()) {
            exhausted = true;
            return false;
        } else {
            index++;
            return true;
        }
    }

    // should this reset exhausted?
    public boolean previousToken() {
        if (index == 0) {
            return false;
        } else {
            index--;
            return true;
        }
    }

    public boolean isExhausted() {
        return exhausted;
    }

    private TokenType determineTokenType(String value) throws TokenizerException {
        for (TokenType tt : tokenizerTypeList) {
            if (tt.isTypeOf(value)) {
                return tt;
            }
        }
        if (value.length() == 0) {
            // if the empty string doesn't have a token type, make one up:
            TokenType emptyType = new ExactToken(EMPTY_NAME, "");
            return emptyType;
        }
        throw new TokenizerException("Token '" + value + "' does not match any known token types!");
    }

    private Token getEmptyToken() {
        TokenType emptyType;
        try {
            emptyType = determineTokenType("");
        } catch (TokenizerException e) {
            // The empty string should always have a token type
            // So this should never get thrown
            throw new RuntimeException(e);
        }
        return new Token("", emptyType);
    }

    public Token currentToken() {
        if (size() == 0) {
            return getEmptyToken();
        } else {
            return tokens.get(index);
        }
    }

    public int getIndex() {
        return index;
    }

    public void setIndex(int newIndex) {
        setIndex(newIndex, false);
    }

    public void setIndex(int newIndex, boolean wasExhausted) {
        if ((size() == 0) && (newIndex == 0)) {
            exhausted = wasExhausted;
        } else if ((newIndex >= 0) && (newIndex < size())) {
            index = newIndex;
            exhausted = wasExhausted;
        } else {
            throw new RuntimeException("index out of bounds");
        }
    }

    public List<Tokenizer> splitTokensOn(Token splitToken) {
        List<Tokenizer> tokenizers = new ArrayList<>();
        List<Token> newTokens = new ArrayList<>();

        for (Token token : tokens) {
            if (splitToken.isSameAs(token)) {
                if (newTokens.size() > 0) {
                    Tokenizer newTokenizer = new Tokenizer(tokenizerTypeList, ignoreWhiteSpace);
                    newTokenizer.resetFromTokenList(newTokens);
                    tokenizers.add(newTokenizer);
                    newTokens = new ArrayList<>();
                }
            } else {
                newTokens.add(token.copy());
            }
        }
        // handle left over tokens
        if (newTokens.size() > 0) {
            Tokenizer newTokenizer = new Tokenizer(tokenizerTypeList, ignoreWhiteSpace);
            newTokenizer.resetFromTokenList(newTokens);
            tokenizers.add(newTokenizer);
        }

        return tokenizers;
    }

    public Token getLastToken() {
        if (size() == 0) {
            return getEmptyToken();
        } else {
            return tokens.get(tokens.size() - 1);
        }
    }

    public TokenizerTypeList getTTL() {
        return tokenizerTypeList;
    }

    public void setTTL(TokenizerTypeList ttl) {
        tokenizerTypeList = ttl;
    }

    public void setIgnoreWhiteSpace(boolean ignoreWS) {
        ignoreWhiteSpace = ignoreWS;
    }

    public void tokenize(String rawText) throws TokenizerException {
        setDefaults();

        while (rawText.length() > 0) {
            boolean hasMatch = false;

            if (ignoreWhiteSpace) {
                Matcher whiteSpaceMatcher = whiteSpacePattern.matcher(rawText);
                if (whiteSpaceMatcher.lookingAt()) {
                    hasMatch = true;
                    rawText = rawText.substring(whiteSpaceMatcher.end());
                }
            }

            if (!hasMatch) {
                for (TokenType tt : tokenizerTypeList) {
                    Pattern ttPattern = tt.getPattern();
                    Matcher ttMatcher = ttPattern.matcher(rawText);
                    if (ttMatcher.lookingAt()) {
                        if (ttMatcher.group().length() > 0) {
                            hasMatch = true;
                            if (!tt.isIgnored()) {
                                if (ttMatcher.groupCount() > 0) {
                                    // concatenate all the "real" groups.
                                    // group 0 is the entire match, and is not counted
                                    // towards groupCount().
                                    StringBuilder sb = new StringBuilder();
                                    for (int i = 1; i < ttMatcher.groupCount() + 1; i++) {
                                        String groupValue = ttMatcher.group(i);
                                        if (groupValue != null && groupValue.length() > 0) {
                                            sb.append(groupValue);
                                        }
                                    }
                                    tokens.add(new Token(sb.toString(), tt, ttMatcher.group(0)));
                                } else {
                                    // no groups to worry about, token value is entire match
                                    tokens.add(new Token(ttMatcher.group(), tt));
                                }
                            }
                            rawText = rawText.substring(ttMatcher.end());
                            break;
                        }
                    }
                }
                if (!hasMatch) {
                    throw new TokenizerException("Beginning of text doesn't match any known TokenTypes: '" + rawText + "'");
                }
            }
        }
    }

}

File: ParserTongue/tokenizer.py
#!/usr/bin/env python3
#tokenizer.py
from errors import *
from tokentype import Token, TokenType
import re

class TokenizerTypeList(object):

    def __init__(self):
        self._typeList = []
        self._indexLookup = {}

    def __len__(self):
        return len(self._typeList)

    def addTokenType(self, tt):
        if not isinstance(tt, TokenType):
            raise TypeError
        if tt.getName() not in self._indexLookup:
            self._indexLookup[tt.getName()] = len(self)
            self._typeList.append(tt)

    def extendWith(self, tts):
        if not isinstance(tts, TokenizerTypeList):
            raise TypeError
        for tt in tts:
            self.addTokenType(tt)

    def __getitem__(self, item):
        if isinstance(item, int):
            return self._typeList[item % len(self)]
        else:
            return self._typeList[self.indexOf(item)]

    def indexOf(self, item):
        if isinstance(item, TokenType):
            return self._indexLookup[item.getName()]
        else:
            return self._indexLookup[item]

    def __iter__(self):
        return iter(self._typeList)

    def __contains__(self, tt):
        return (tt.getName() in self._indexLookup)

    def __str__(self):
        value = "TokenizerTypeList: ["
        for tt in self:
            value += str(tt) + ", "
        value = value[:-2] # get rid of trailing ', '
        value += "]"
        return value

def defaultGrammarTTL():
    gTTL = TokenizerTypeList()
    gTTL.addTokenType(TokenType("End", r';'))
    gTTL.addTokenType(TokenType("Define", r'='))
    gTTL.addTokenType(TokenType("External", r':'))
    gTTL.addTokenType(TokenType("RegEx", r'~'))
    gTTL.addTokenType(TokenType("Control", r'[()[\]{}|,]'))
    gTTL.addTokenType(TokenType("Identifier", r'[a-zA-Z][a-zA-Z0-9_]*'))
    gTTL.addTokenType(TokenType("Terminal", r'\'([^\']*)\'|"([^"]*)"'))
    gTTL.addTokenType(TokenType("Comment", r'(#.*\n)|(#.*\Z)', True))
    return gTTL

def getTTLForAlphabet(alphabet):
    alphTTL = TokenizerTypeList()
    for letter in alphabet:
        alphTTL.addTokenType(TokenType(letter, letter))
    return alphTTL

class Tokenizer(object):

    def __init__(self, ttl = defaultGrammarTTL(), ignoreWhiteSpace = True):
        self.tokens = []
        self.index = 0
        self._exhausted = False
        self.ttl = ttl
        self._ignoreWS = ignoreWhiteSpace
        self._wsRE = re.compile(r'\s+')

    def nextToken(self):
        self.index += 1
        if self.index >= len(self.tokens):
            self._exhausted = True
            if len(self) > 0:
                self.index = len(self.tokens)-1
            else:
                self.index = 0
            return False
        return True

    def previousToken(self):
        self.index -= 1
        if self.index < 0:
            self.index = 0
            return False
        return True

    def isExhausted(self):
        return self._exhausted

    def currentToken(self):
        if len(self) == 0:
            return self.getEmptyToken()
        else:
            return self.tokens[self.index]

    def getEmptyToken(self):
        emptyType = self.determineTokenType('')
        return Token('',emptyType)

    def getIndex(self):
        return self.index

    def setIndex(self, index, wasExhausted = False):
        if len(self) == 0 and index == 0:
            self._exhausted = wasExhausted
        elif index >= 0 and index < len(self):
            self.index = index
            self._exhausted = wasExhausted
        else:
            raise IndexError

    def __len__(self):
        return len(self.tokens)

    def __str__(self):
        return str(self.tokens) + ';index=' + str(self.index) + ';exhausted?' + str(self._exhausted)

    def splitTokensOn(self, splitToken):
        tokenizers = []
        newTokens = []
        for t in self.tokens:
            if t == splitToken:
                if len(newTokens) > 0:
                    newTokenizer = Tokenizer(self.ttl, self._ignoreWS)
                    newTokenizer.resetFromTokenList(newTokens)
                    tokenizers.append(newTokenizer)
                    newTokens = []
            else:
                newTokens.append(t)
        if len(newTokens) > 0:
            newTokenizer = Tokenizer(self.ttl, self._ignoreWS)
            tokenizers.append(newTokenizer.resetFromTokenList(newTokens))
        return tokenizers

    def resetFromTokenList(self, tokenList):
        self.tokens = []
        for t in tokenList:
            if not isinstance(t, Token):
                raise TokenizerCreationError("in Tokenizer.resetFromTokenList : given list contains non-token!")
            self.tokens.append(t)

    # used by Grammar to check the last token is ';' before splitting.
    def getLastToken(self):
        return self.tokens[-1]

    # Not super recomended, but provided anyway
    def getTokenList(self):
        return self.tokens

    def setTTL(self, ttl):
        self.ttl = ttl

    def getTTL(self):
        return self.ttl

    def setIgnoreWhiteSpace(self, value):
        self._ignoreWS = bool(value)

    def determineTokenType(self, tokenStr):
        for tokenType in self.ttl:
            if tokenType.isTypeOf(tokenStr):
                return tokenType
        raise UnknownTokenTypeError("Token '" + tokenStr + "' does not match any known tokenTypes!")

    def tokenize(self, rawText):
        self.tokens = []
        self.index = 0
        self._exhausted = False
        text = rawText

        while len(text) > 0:
            matchObj = None
            if self._ignoreWS:
                matchObj = self._wsRE.match(text)
            if matchObj is None:
                for tt in self.ttl:
                    ttRE = tt.getRE()
                    matchObj = ttRE.match(text)
                    if matchObj is not None:
                        if matchObj.end() > 0:
                            break
                if matchObj is not None:
                    groups = matchObj.groups()
                    matchText = matchObj.group(0)
                    if len(groups) > 0: # Concatenate groups together.
                        sumtext = ''
                        for g in groups:
                            if g is not None:
                                sumtext += str(g)
                        matchText = sumtext
                    if not tt.isIgnored():
                        self.tokens.append(Token(matchText, tt, matchObj.group(0)))
                    text = text[matchObj.end():]
                else:
                    raise TokenizerNoMatchError("Beginning of text doesn't match any known TokenTypes: '" + text + "'")
            else:
                text = text[matchObj.end():]


